pkg/log/log.go

```diff
@@ -2,177 +2,162 @@ package log
-	// be added before calling pflag.Parse().
-	pflag.CommandLine.AddFlagSet(zapFlagSet)
-
-	// Add flags registered by imported packages (e.g. glog and
-	// controller-runtime)
-	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
-
-	pflag.Parse()
-
-	// Use a zap logr.Logger implementation. If none of the zap
-	// flags are configured (or if the zap flag set is not being
-	// used), this defaults to a production zap logger.
-	//
-	// The logger instantiated here can be changed to any logger
-	// implementing the logr.Logger interface. This logger will
-	// be propagated through the whole operator, generating
-	// uniform and structured logs.
-	logf.SetLogger(zap.Logger())
-
-	// set the logger.
-	log = Logger{
-		name:          withName,
-		lgr:           logf.Log.WithName(withName),
-		out:           os.Stdout,
-		isTestingMode: false,
+// Init initializes the logger.
+func Init(withName string) {
+	once.Do(func() {
+		logger = newLogger(withName)
+	})
+}
+
+func newLogger(withName string) *Logger {
+	return &Logger{
+		logr: logf.Log.WithName(withName),
+		name: withName,
 	}
+}
 
-	return &log
+// ZapLoggerTo returns a new Logger implementation using Zap which logs
+// to the given destination, instead of stderr.
+func ZapLoggerTo(out io.Writer, development bool) *Logger {
+	nl := newLogger(logger.name)
+	nl.logr = logf.ZapLoggerTo(out, development).WithName(logger.name)
+	return nl
 }
 
-func (p *Logger) SetOutput(out io.Writer, isTestingMode bool) *Logger {
-	// WithValues, WithName and ZapLoggerTo all result in a new logger instance.
-	// The values stored in the Logger struct must be set again.
-	if len(log.tags) > 0 {
-		log.lgr = logf.ZapLoggerTo(out, isTestingMode).WithName(log.name).WithValues(log.tags...)
-	} else {
-		log.lgr = logf.ZapLoggerTo(out, isTestingMode).WithName(log.name)
-	}
+// Info logs a non-error message.
+func Info(ctx *gin.Context, msg string) {
+	logger.Info(ctx, msg)
+}
+
+// Infof logs a non-error formatted message.
+func Infof(ctx *gin.Context, msg string, args ...string) {
+	logger.Infof(ctx, msg, args...)
+}
 
-	log.out = out
-	log.isTestingMode = isTestingMode
-	return &log
+// Error logs the error with the given message.
+func Error(ctx *gin.Context, err error, msg string) {
+	logger.Error(ctx, err, msg)
 }
 
-// GetLogger returns the current logger object.
-func GetLogger() *Logger {
-	return &log
+// Errorf logs the error with the given formatted message.
+func Errorf(ctx *gin.Context, err error, msg string, args ...string) {
+	logger.Errorf(ctx, err, msg, args...)
 }
 
-// Info logs are used for non-error messages. It will log a message with
-// the given key/value pairs as context.
-func (p *Logger) Info(ctx *gin.Context, msg string) *Logger {
-	return p.Infof(ctx, msg)
+// WithValues creates a new logger with additional key-value pairs in the context
+func WithValues(keysAndValues map[string]interface{}) *Logger {
+	return logger.WithValues(keysAndValues)
```

Comment:
WithValues returns a new logger instance. if we do this the `WithName` value which we set earlier will be eliminated along with any other WithValues and the output...

Reply:
The name of the logger is properly propagated by the underling logger.
`log.WithValues(...).Infof(...)` logs the message with `"logger":"reg-service"`.

Reply:
hmm... from my testing, i noticed that the output was not set and it will panic due to a nil pointer reference. also from testing i noticed that the withName field disappears... i tried just implementing the info func and i see the same results. perhaps i am doing something wrong :S but the tests fail